arm64: reset: flush D-Cache before it is disabled

JiafeiPan opened the pull request:
In the commit 573a712bed06d984aab4a0b7424fc146e72c0921 patch "arm64: reset: disable cache and MMU for safety", it disables D-Cache and MMU for safety, but in some cases, for example the code is loaded into memory by hardware debugger, we need to flush D-Cache before disable it in order to make sure the data is coherent in the system, otherwise it will report "Synchronous Abort" when D-Cache is disabled.

So in this PR need to revert commit 867ba172b46ccc528f2f9f1d3fb532e50e30ceb4 "arch: arm64: cache: delete arm64_dcache_all" in order to flush all d-cache, although it can't broadcast the flush operation to other CPU Cores, but it is safety and depend on hardware to make sure it is safe for all CPU Cores. Even for SMP OS, one thread want to flush all the Cores, there is no software notification to other Cores, it up to system bus to make sure it is safe.

JiafeiPan committed:
Revert "arch: arm64: cache: delete arm64_dcache_all"

In order to clean and invalidate all D-Cache, so reverts commit:
867ba172b46ccc528f2f9f1d3fb532e50e30ceb4.

Although no software broadcast between different CPU Cores for  D-Cache
maintaining operation, but the hardware bus system should make sure it
is safety for all CPU Cores, because it is common for other SMP OS.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
arm64: reset: flush D-Cache before it is disabled

In the commit 573a712bed06d984aab4a0b7424fc146e72c0921 patch "arm64:
reset: disable cache and MMU for safety", it disables D-Cache and MMU
for safety, but in some cases, for example the code is loaded into memory
by hardware debugger, we need to flush D-Cache before disable it in
order to make sure the data is coherent in the system, otherwise it
will report "Synchronous Abort" when D-Cache is disabled.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan commented:
Hi, @carlocaione , would you please review this PR? thanks.

ithinuel commented on arch/arm64/core/reset.S:
```diff
@@ -60,7 +60,7 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__reset_prep_c)
 2:
 	/* Disable alignment fault checking, disable D-Cache, I-Cache, MMU for safety */
 	mrs	x0, sctlr_el2
-	mov	x1, #(SCTLR_A_BIT | SCTLR_C_BIT | SCTLR_M_BIT | SCTLR_I_BIT)
```
This sort of reverts 573a712bed06d984aab4a0b7424fc146e72c0921 but keeps less efficient code (2 instr instead of 1) and gets the comment 2lines above out of sync.

I think you can revert the commit completely to get back to the previous state of the instrs & comment, to then add your alternative solution in reset.c.

JiafeiPan replied:
Revert the old one, thanks.

ithinuel commented on include/zephyr/arch/arm64/cache.h:
The rational for the removal of `arm64_dcache_all` in 867ba172b46ccc528f2f9f1d3fb532e50e30ceb4 seem sensible.
Since this PR reverts it, how do you address these concerns here ?

cc @xakep-amatop may be able to advise ?

JiafeiPan replied:
Hi, @ithinuel , in this PR need to revert commit https://github.com/zephyrproject-rtos/zephyr/commit/867ba172b46ccc528f2f9f1d3fb532e50e30ceb4 "arch: arm64: cache: delete arm64_dcache_all" in order to flush all d-cache, although it can't broadcast the flush operation to other CPU Cores, but it is safety and depend on hardware to make sure it is safe for all CPU Cores. Even for SMP OS, one thread want to flush all the Cores, there is no software notification to other Cores, it up to system bus to make sure it is safe.

JiafeiPan committed:
Revert "arm64: reset: disable cache and MMU for safety"

This reverts commit 573a712bed06d984aab4a0b7424fc146e72c0921.

The original patch disable cache and MMU, but in some cases, for example
the code is loaded into memory by hardware debugger, we need to flush
D-Cache before disable it in order to make sure the data is coherent
in the system, otherwise it will report "Synchronous Abort" when D-Cache
is disabled.

This patch revert the old one and the following patch will flush
the cache before disable it.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan commented:
Rebased.

JiafeiPan commented:
@ithinuel , may I know whether your requested change has been address? thanks.

ithinuel reviewed (commented):
I'm happy with the change sequence (in it's form). 👍
I don't know much about the impl détails so a review from an expert in the field would still be required.

JiafeiPan committed:
arch: arm64: cache: optionally enable APIs for operation all dcache

Added new configuration item to optionally enable APIs for operation
all data cache, by default these APIs are disabled.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
soc: imx: disable dcache until mmu is enabled during booting

Enable CONFIG_ARM64_BOOT_DISABLE_DCACHE for i.MX Cortex-A platforms.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan commented:
v2: 
1. Add new configuration item CONFIG_ARM64_DCACHE_ALL_API to optionally enable data cache APIs to operate all data cache, it is disabled by default, so no impaction to currently all the other platforms.
2. Add new configuration item CONFIG_ARM64_BOOT_DISABLE_DCACHE to optionally disable data cache if it is enabled during booting from EL2, it is disabled by default, so no impaction to currently all the other platforms.
3. Enabled CONFIG_ARM64_BOOT_DISABLE_DCACHE on all i.MX A-Core platforms.
4. rebased.

wearyzen commented on arch/arm64/core/Kconfig:
```diff
@@ -357,4 +357,14 @@ config MAX_XLAT_TABLES
+if CACHE_MANAGEMENT
+
+config ARM64_DCACHE_ALL_API
+	bool "Provde data cache APIs to operate all data caches"
```
Instead of the if CACHE_* please use 
`depends on CACHE_MANAGEMENT && DCACHE`

JiafeiPan replied:
Done, thanks.

wearyzen commented on arch/arm64/core/Kconfig:
`Added new configuration item to optinally enable APIs for operation
all data cache, by default these APIs are disabled.`
nit: **optionally**

JiafeiPan replied:
Fixed, thanks.

wearyzen commented on arch/arm64/core/Kconfig:
```diff
@@ -365,6 +365,15 @@ config ARM64_DCACHE_ALL_API
 	  Enable this option to provide the data cache APIs to flush or
 	  invalidate all data caches.
 
+config ARM64_BOOT_DISABLE_DCACHE
```
`but in some cases, for example the code is loaded into memory
by hardware debugger, we need to flush D-Cache before disable it in
order to make sure the data is coherent in the system, otherwise it
will report "Synchronous Abort" when D-Cache is disabled.`
- can we use an existing configs e.g. CONFIG_DEBUG_/CONFIG_DEBUG_THREAD_INFO to differentiate this behavior?

JiafeiPan replied:
I am not sure you mean to print some debug information when ARM64_BOOT_DISABLE_DCACHE is enabled? but this code locates at very early during booting, printk is still not ready.

wearyzen replied:
I mean instead of adding a new config option can we use the existing flag [here](https://github.com/zephyrproject-rtos/zephyr/pull/88415/commits/d6eb8d56aa1c312d74fbe84f14886415ea202390#diff-8b11b0d20afcf442528968d8c30f90c6a448886e4acfc87e39db32a5358e6841R126)
something like :
```
#ifndef CONDIF_DEBUG
#ifdef CONFIG_ARM64_BOOT_DISABLE_DCACHE
	/* Disable D-Cache if it is enabled, it will re-enabled when MMU is enabled at EL1 */
	if (reg & SCTLR_C_BIT) {
		/* Clean and invalidate dcache before it is disabled to make sure memory to be
		 * coherent
		 */
		arch_dcache_flush_and_invd_all();
		barrier_isync_fence_full();
		/* Disable D-Cache and MMU for EL2 */
		reg &= ~(SCTLR_C_BIT | SCTLR_M_BIT);
		write_sctlr_el2(reg);
		/* Invalidate TLB entries */
		__asm__ volatile("dsb ishst; tlbi alle2; dsb ish; isb" : : : "memory");
	}
#endif 
```

either with CONFIG_DEBUG or other config which is enabled when a debugger is used.

JiafeiPan replied:
Not only use debugger need these code, there are some other scenario such as using go command in uboot to boot zephyr, as MMU and dcache has already enabled in uboot, so it will also have this issue, so I think we still need a new config item to enable this feature.

JiafeiPan replied:
Hi, @wearyzen , would you please review this PR again? thanks.

wearyzen replied:
nit: DO you think ARM64_EARLY_DISABLE_DCACHE or ARM64_EL2_DISABLE_DCACHE might be a better name?

JiafeiPan replied:
I prefer to use ARM64_BOOT_DISABLE_DCACHE, actually we only disable dcache only during booting process, I think need to avoid the misunderstanding that we diable dcache after zephyr boot up.

JiafeiPan commented on arch/arm64/core/Kconfig:
```diff
@@ -357,4 +357,14 @@ config MAX_XLAT_TABLES
 
+if CACHE_MANAGEMENT
+
+config ARM64_DCACHE_ALL_API
```
Updated to use CONFIG_ARM64_DCACHE_ALL_OPS, thanks.

wearyzen commented on arch/arm64/core/Kconfig:
```diff
@@ -364,4 +364,13 @@ config ARM64_DCACHE_ALL_OPS
+	help
+	  To make it safe, if data cache is enabled in case of Zephyr is booting
+	  from EL2, enable this option, it will clean and invalidate all data
+	  cache and then disable datat cache, it will will be re-enabled after
```
*data

JiafeiPan replied:
fixed, thanks.

JiafeiPan commented on arch/arm64/core/reset.c:
```diff
@@ -122,6 +123,21 @@ void z_arm64_el2_init(void)
+#ifdef CONFIG_ARM64_BOOT_DISABLE_DCACHE
+	/* Disable D-Cache if it is enabled, it will re-enabled when MMU is enabled at EL1 */
+	if (reg & SCTLR_C_BIT) {
+		/* Clean and invalidate dcache before it is disabled to make sure memory to be
```
fixed, thanks.

JiafeiPan commented:
Rebased and addressed @wearyzen 's comments, thanks.
